crates/driver/src/domain/competition/solution/fee.rs

```diff
@@ -138,33 +145,58 @@ impl Fulfillment {
         factor: f64,
     ) -> Result<eth::U256, Error> {
         let surplus = self.surplus_over_reference_price(sell_amount, buy_amount, prices)?;
-        let surplus_in_sell_token = self.surplus_in_sell_token(surplus, prices)?;
-        apply_factor(surplus_in_sell_token, factor)
+        apply_factor(surplus, factor)
     }
 
+    /// Computes the volume based fee in surplus token
+    ///
+    /// The volume is defined as a full sell amount (including fees) conversion
+    /// to the full buy amount (user point of view)
     fn fee_from_volume(&self, prices: ClearingPrices, factor: f64) -> Result<eth::U256, Error> {
         let executed = self.executed().0;
-        let executed_sell_amount = match self.order().side {
+        let executed_in_surplus_token = match self.order().side {
             Side::Buy => {
                 // How much `sell_token` we need to sell to buy `executed` amount of `buy_token`
                 executed
                     .checked_mul(prices.buy)
                     .ok_or(trade::Error::Overflow)?
                     .checked_div(prices.sell)
                     .ok_or(trade::Error::DivisionByZero)?
+                    .checked_add(
+                        // surplus_fee is always expressed in sell token
+                        self.surplus_fee()
+                            .map(|fee| fee.0)
+                            .ok_or(trade::Error::ProtocolFeeOnStaticOrder)?,
+                    )
+                    .ok_or(trade::Error::Overflow)?
             }
-            Side::Sell => executed,
+            Side::Sell => {
+                // How much `buy_token` we get when we sell `executed` amount of `sell_token`
+                executed
+                    .checked_mul(prices.sell)
+                    .ok_or(trade::Error::Overflow)?
+                    .checked_div(prices.buy)
+                    .ok_or(trade::Error::DivisionByZero)?
+            }
+        };
+        apply_factor(executed_in_surplus_token, factor)
+    }
+
+    /// Returns the fee denominated in the sell token.
+    pub fn in_sell_token(
```

Comment:
> maybe it could make sense to expose

But the function also requires `Side` so putting it on Asset would be overkill.

I created the `protocol_fee_in_sell_token` instead. Is that better IYO or same?

Reply:
You wouldn't need side if all functions returned `eth:Asset`. Then 

```rs
let protocol_fee = self.protocol_fee_in_sell_token(prices)?;
```

becomes

```rs
let protocol_fee = self.protocol_fee().in_token(self.order().sell.token, prices)
```

(for buy order protocol fee will be already in sell so `in_token` will multiply with 1, for sell orders it will convert into sell token.

✨

But I'm also ok leaving this as is and leave this refactoring up for another time™️